src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertForEachToLinqQueryTests.cs

```diff
@@ -0,0 +1,2842 @@
+        public async Task ToListToNewArrayElement()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums, List<int>[] lists)
+    {
+        lists[0] = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                lists[0].Add(n1);
+            }
+        }|]
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    void M(IEnumerable<int> nums, List<int>[] lists)
+    {
+        lists[0] = (from int n1 in nums
+                    from int n2 in nums
+                    select n1).ToList();
+    }
+}
+";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListHashSetNoConversion()
+        {
+            string source = @"
+using System.Collections.Generic;
+class C
+{
+    void M(IEnumerable<int> nums)
+    {
+        var hashSet = new HashSet<int>();
+        [|foreach (int n1 in nums)
+        {
+            hashSet.Add(n1);
+        }|]
+    }
+}
+";
+            await TestMissingAsync(source);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListMergeWithReturn()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        var list = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                list.Add(n1);
+            }
+        }|]
+
+        return list;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        return (from int n1 in nums
+                from int n2 in nums
+                select n1).ToList();
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListSeparateDeclarationAndAssignmentMergeWithReturn()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        List<int> list;
+        list = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                list.Add(n1);
+            }
+        }|]
+
+        return list;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        List<int> list;
```

Comment:
only if we could change refactorings.. we could have run remove unused locals...

Reply:
I like the idea. We should be able to call other refactorings as post-actions to existing ones. Thank you!

---
In reply to: [195856741](https://github.com/dotnet/roslyn/pull/27130#discussion_r195856741) [](ancestors = 195856741)